wazo_confd/plugins/email/schema.py

```diff
@@ -1,54 +1,53 @@
 # Copyright 2021 The Wazo Authors  (see the AUTHORS file)
 # SPDX-License-Identifier: GPL-3.0-or-later
 
-from marshmallow import fields, post_load, pre_dump
+from marshmallow import fields, post_load, pre_dump, ValidationError
 from marshmallow.validate import Length
 
 from wazo_confd.helpers.mallow import BaseSchema
 
 
 class _RewriteRule(BaseSchema):
-    match = fields.String()
-    replacement = fields.String()
+    match = fields.String(required=True)
+    replacement = fields.String(required=True)
+
+    @pre_dump
+    def from_canonical(self, data, **kwargs):
+        if data and ' ' not in data:
+            raise ValidationError(
+                f'Data value "{data}" should be of format "match replacement"'
+            )
+        match, replacement = data.split(maxsplit=1)
+        return {'match': match, 'replacement': replacement}
+
+    @post_load
+    def to_canonical(self, data, **kwargs):
+        return f"{data['match']} {data['replacement']}"
 
 
 class EmailConfigSchema(BaseSchema):
-    domain_name = fields.String(validate=Length(max=255), missing='')
-    from_ = fields.String(data_key='from', validate=Length(max=255), missing='')
-    address_rewriting_rules = fields.List(fields.Nested(_RewriteRule), missing=[])
-    smtp_host = fields.String(validate=Length(max=255), missing='')
-    fallback_smtp_host = fields.String(validate=Length(max=255), missing='')
-
-    def _rewriting_rules_from_db(self, canonical):
-        rules = []
-        lines = [line for line in canonical.split('\\n') if line]
-        for line in lines:
-            match, replacement = line.split(maxsplit=1)
-            rules.append({'match': match, 'replacement': replacement})
-        return rules
-
-    def _rewriting_rules_to_db(self, rules):
-        lines = [f"{rule['match']} {rule['replacement']}" for rule in rules]
-        return '\\n'.join(lines)
+    domain_name = fields.String(
+        attribute='mydomain', validate=Length(max=255), missing=''
+    )
+    from_ = fields.String(
+        attribute='origin', data_key='from', validate=Length(max=255), missing=''
+    )
+    address_rewriting_rules = fields.List(
+        fields.Nested(_RewriteRule, missing=None), attribute='canonical', missing=[]
+    )
+    smtp_host = fields.String(
+        attribute='relayhost', validate=Length(max=255), missing=''
+    )
+    fallback_smtp_host = fields.String(
+        attribute='fallback_relayhost', validate=Length(max=255), missing=''
+    )
 
     @pre_dump
-    def from_db_model(self, data, **kwargs):
-        result = {
-            'domain_name': data.mydomain,
-            'from_': data.origin,
-            'address_rewriting_rules': self._rewriting_rules_from_db(data.canonical),
-            'smtp_host': data.relayhost,
-            'fallback_smtp_host': data.fallback_relayhost,
-        }
-        return result
+    def split_canonical(self, data, **kwargs):
+        data.canonical = [line for line in data.canonical.split('\\n') if line]
```

Comment:
Oh this is very dangerous, you change the data object, but this object is a sqlalchemy object and if it's still bound to the session, it will try to change the value. You should use an attribute that doesn't exist as column